### PR TITLE
feat: update context to take time

### PR DESCRIPTION
**Pull request by fchikwekwe**
Description: Update contexts so that time related setters and getters support time objects.

Link to tracking Issue: #22010

Testing: unit tests for each new setter/getter

Documentation: READMEs for `ottldatapoint`, `ottllog`, `ottlspan`, `ottlspanevent`.

### CODE REVIEW

File: pkg/ottl/contexts/ottldatapoint/datapoint_test.go [hunk 1]
```diff
@@ -108,14 +108,27 @@ func Test_newPathGetSetter_NumberDataPoint(t *testing.T) {
 		valueType pmetric.NumberDataPointValueType
 	}{
 		{
-			name: "start_time_unix_nano",
+			name: "start_time",
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply pkg/ottl/contexts/ottldatapoint/datapoint_test.go:124]: Yes, good suggestion.

File: pkg/ottl/contexts/ottldatapoint/datapoint_test.go [hunk 2]
```diff
@@ -108,14 +108,27 @@ func Test_newPathGetSetter_NumberDataPoint(t *testing.T) {
...
-			orig:   int64(100_000_000),
-			newVal: int64(200_000_000),
+			orig:   time.Date(1970, 1, 1, 0, 0, 0, 100000000, time.UTC),
+			newVal: time.Date(1970, 1, 1, 0, 0, 0, 86400000000000, time.UTC),
+			modified: func(datapoint pmetric.NumberDataPoint) {
+				datapoint.SetStartTimestamp(pcommon.NewTimestampFromTime(time.Unix(86400, 0)))
+			},
+		},
+		{
+			name: "time",
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply pkg/ottl/contexts/ottldatapoint/datapoint_test.go:137]: There still is a `time_unix_nano` case.

File: pkg/ottl/contexts/ottldatapoint/datapoint_test.go [hunk 3]
```diff
@@ -108,14 +108,27 @@ func Test_newPathGetSetter_NumberDataPoint(t *testing.T) {
...
+			path: []ottl.Field{
+				{
+					Name: "time",
+				},
+			},
+			orig:   time.Date(1970, 1, 1, 0, 0, 0, 500000000, time.UTC),
+			newVal: time.Date(1970, 1, 1, 0, 0, 0, 200000000, time.UTC),
 			modified: func(datapoint pmetric.NumberDataPoint) {
 				datapoint.SetStartTimestamp(pcommon.NewTimestampFromTime(time.UnixMilli(200)))
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply pkg/ottl/contexts/ottldatapoint/datapoint_test.go]: Yes, good catch. I updated this.

File: pkg/ottl/contexts/ottllog/log_test.go [hunk 4]
```diff
@@ -83,6 +96,19 @@ func Test_newPathGetSetter(t *testing.T) {
...
+			name: "observed time",
+			path: []ottl.Field{
+				{
+					Name: "observed_time",
+				},
+			},
+			orig:   time.Date(1970, 1, 1, 0, 0, 0, 200000000, time.UTC),
+			newVal: time.Date(1970, 1, 1, 0, 0, 0, 100000000, time.UTC),
+			modified: func(log plog.LogRecord, il pcommon.InstrumentationScope, resource pcommon.Resource, cache pcommon.Map) {
+				log.SetTimestamp(pcommon.NewTimestampFromTime(time.UnixMilli(200)))
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply pkg/ottl/contexts/ottllog/log_test.go]: Updated. ty.

File: pkg/ottl/contexts/ottllog/log_test.go [hunk 5]
```diff
@@ -83,6 +96,19 @@ func Test_newPathGetSetter(t *testing.T) {
...
 		},
+		{
+			name: "observed time",
+			path: []ottl.Field{
+				{
+					Name: "observed_time",
+				},
+			},
+			orig:   time.Date(1970, 1, 1, 0, 0, 0, 200000000, time.UTC),
+			newVal: time.Date(1970, 1, 1, 0, 0, 0, 100000000, time.UTC),
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply pkg/ottl/contexts/ottllog/log_test.go]: yes, updated.

[the remark this answers is not in the document]
Comment (fchikwekwe) [reply pkg/ottl/contexts/ottllog/log_test.go]: and thank you, I was fiddling with this one for a while and for some reason could not figure the error out.

File: pkg/ottl/contexts/ottllog/log_test.go [hunk 6]
```diff
@@ -83,6 +96,19 @@ func Test_newPathGetSetter(t *testing.T) {
...
 			},
 		},
+		{
+			name: "observed time",
+			path: []ottl.Field{
+				{
+					Name: "observed_time",
+				},
+			},
+			orig:   time.Date(1970, 1, 1, 0, 0, 0, 200000000, time.UTC),
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply pkg/ottl/contexts/ottllog/log_test.go]: Done!

File: pkg/ottl/contexts/internal/span.go
Comment (TylerHelmuth) [inline pkg/ottl/contexts/internal/span.go:1]: Please add new tests to `contexts/internal/span_test.go`.

File: pkg/ottl/contexts/ottldatapoint/datapoint_test.go [hunk 7]
```diff
@@ -111,13 +111,39 @@ func Test_newPathGetSetter_NumberDataPoint(t *testing.T) {
...
 			},
-			orig:   int64(100_000_000),
-			newVal: int64(200_000_000),
+			orig:   time.Date(1970, 1, 1, 0, 0, 0, 100000000, time.UTC),
+			newVal: time.Date(1970, 1, 1, 0, 0, 0, 86400000000000, time.UTC),
 			modified: func(datapoint pmetric.NumberDataPoint) {
-				datapoint.SetStartTimestamp(pcommon.NewTimestampFromTime(time.UnixMilli(200)))
+				datapoint.SetStartTimestamp(pcommon.NewTimestampFromTime(time.Unix(86400, 0)))
+			},
+		},
```
Comment (TylerHelmuth) [inline pkg/ottl/contexts/ottldatapoint/datapoint_test.go:122]: We shouldn't need to change any tests for the `*_unix_nano` accessors.

File: pkg/ottl/contexts/ottldatapoint/datapoint_test.go [hunk 8]
```diff
@@ -111,7 +111,7 @@ func Test_newPathGetSetter_NumberDataPoint(t *testing.T) {
 			name: "start_time_unix_nano",
 			path: []ottl.Field{
 				{
-					Name: "start_time_unix_nano",
+					Name: "start_time",
```
Comment (TylerHelmuth) [inline pkg/ottl/contexts/ottldatapoint/datapoint_test.go:114]:
```suggestion
					Name: "start_time_unix_nano",
```

**Review by TylerHelmuth — APPROVED** [review]
LGTM

**Comment by TylerHelmuth** [comment]
@evan-bradley this is the last of many PRs to support interacting with time fields via time.Time, have any thoughts?

**Review by evan-bradley — APPROVED** [review]
Looks good to me. Thanks for all the improvements to OTTL's time support, @fchikwekwe!

**Pull request merged**